src/NServiceBus.Core/Transports/Msmq/ReceiveWithTransactionScope.cs

```diff
@@ -19,45 +19,118 @@ public ReceiveWithTransactionScope(TransactionOptions transactionOptions)
 
         public override async Task ReceiveMessage(MessageQueue inputQueue, MessageQueue errorQueue, CancellationTokenSource cancellationTokenSource, Func<PushContext, Task> onMessage)
         {
-            using (var scope = new TransactionScope(TransactionScopeOption.Required, transactionOptions, TransactionScopeAsyncFlowOption.Enabled))
-            {
-                var message = inputQueue.Receive(TimeSpan.FromMilliseconds(10), MessageQueueTransactionType.Automatic);
+            Func<ErrorContext, Task<RecoveryAction>> onError = context => Task.FromResult(new RecoveryAction());
 
-                Dictionary<string, string> headers;
+            Dictionary<string, string> headers = null;
+            Message message = null;
 
-                try
-                {
-                    headers = MsmqUtilities.ExtractHeaders(message);
-                }
-                catch (Exception ex)
+            try
+            {
+                using (var scope = new TransactionScope(TransactionScopeOption.Required, transactionOptions, TransactionScopeAsyncFlowOption.Enabled))
                 {
-                    var error = $"Message '{message.Id}' is corrupt and will be moved to '{errorQueue.QueueName}'";
-                    Logger.Error(error, ex);
+                    message = inputQueue.Receive(TimeSpan.FromMilliseconds(10), MessageQueueTransactionType.Automatic);
 
-                    errorQueue.Send(message, MessageQueueTransactionType.Automatic);
+                    if (PerformRecoverabilityAction(message))
```

Comment:
Since we have to rollback in this "mode" we check if we need to execute any recoverability actions before we proceed with calling the core pipeline. This would not be needed for transports like rabbit that don't have to rollback